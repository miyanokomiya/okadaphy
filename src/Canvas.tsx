import * as React from 'react'
import PropTypes from 'prop-types'
import App from './app'
import { IOptions } from '../types'

type Props = IOptions & {
  width?: number
  height?: number
  count?: number
}

const Canvas: React.FC<Props> = props => {
  const canvasRef = React.useRef(null)

  React.useEffect(() => {
    const app = new App({ canvas: canvasRef.current as any })
    app.importFromString({
      fillStyle: props.fillStyle,
      strokeStyle: props.strokeStyle,
      text: props.text,
    })
    app.setGravity(props.gravityX, props.gravityY)
    app.run()
    return () => app.dispose()
  }, [props])

  return <canvas ref={canvasRef} width={props.width} height={props.height} />
}

Canvas.propTypes = {
  width: PropTypes.number,
  height: PropTypes.number,
  text: PropTypes.string.isRequired,
  fillStyle: PropTypes.string.isRequired,
  strokeStyle: PropTypes.string.isRequired,
  gravityX: PropTypes.number.isRequired,
  gravityY: PropTypes.number.isRequired,
  count: PropTypes.number,
}
Canvas.defaultProps = {
  width: 300,
  height: 300,
  count: 0,
}

export default Canvas
