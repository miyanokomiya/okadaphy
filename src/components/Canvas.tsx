import * as React from 'react'
import PropTypes from 'prop-types'
import Grid from '@material-ui/core/Grid'
import Button from '@material-ui/core/Button'
import App from '../app'
import { IConfig } from '../../types'

export type Props = {
  width?: number
  height?: number
  config: IConfig
  count?: number
}

const Canvas: React.FC<Props> = props => {
  const canvasRef = React.useRef(null)
  const [app, setApp] = React.useState<App | null>(null)
  const [running, setRunning] = React.useState(true)

  // app初期化と片付け
  React.useEffect(() => {
    const app = new App({ canvas: canvasRef.current as any })
    app.run()
    setApp(app)
    setRunning(true)
    return () => app.dispose()
  }, [])

  // テキストインポート
  React.useEffect(() => {
    if (!app) return
    app.clear()
    app.importFromString(props.config.text)
  }, [app, props.config.text, props.count])

  // プロパティ反映
  React.useEffect(() => {
    if (!app) return
    app.setGravity(props.config.gravityX, props.config.gravityY)
    app.setStyle({ fillStyle: props.config.fillStyle, strokeStyle: props.config.strokeStyle })
  }, [
    app,
    props.config.gravityX,
    props.config.gravityY,
    props.config.fillStyle,
    props.config.strokeStyle,
  ])

  const onClickPlay = React.useCallback(() => {
    if (!app) return
    running ? app.stop() : app.run()
    setRunning(!running)
  }, [app, running])

  const onClickStep = React.useCallback(() => {
    if (!app) return
    app.step()
  }, [app])

  return (
    <div>
      <canvas ref={canvasRef} width={props.width} height={props.height} />
      <Grid container spacing={2}>
        <Grid item>
          <Button
            variant="contained"
            color={running ? 'secondary' : 'primary'}
            onClick={onClickPlay}
          >
            {running ? 'Pause' : 'Start'}
          </Button>
        </Grid>
        <Grid item>
          <Button variant="contained" onClick={onClickStep}>
            Step
          </Button>
        </Grid>
      </Grid>
    </div>
  )
}

Canvas.propTypes = {
  width: PropTypes.number,
  height: PropTypes.number,
  config: PropTypes.shape({
    text: PropTypes.string.isRequired,
    fillStyle: PropTypes.string.isRequired,
    strokeStyle: PropTypes.string.isRequired,
    gravityX: PropTypes.number.isRequired,
    gravityY: PropTypes.number.isRequired,
  }).isRequired,
  count: PropTypes.number,
}
Canvas.defaultProps = {
  width: 300,
  height: 300,
  count: 0,
}

export default Canvas
