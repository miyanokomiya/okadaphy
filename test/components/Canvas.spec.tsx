import * as React from 'react'
import renderer from 'react-test-renderer'
import Canvas, { Props } from '../../src/components/Canvas'
import { mockConfig } from '../mock'

jest.mock('@material-ui/core/Slider', () => 'Slider')
jest.mock('@material-ui/core/TextField', () => 'TextField')
jest.mock('react-color', () => ({ CompactPicker: 'CompactPicker' }))

describe('src/components/ConfigForm', () => {
  it('snapshot', () => {
    const props: Props = {
      config: mockConfig(),
      width: 100,
      height: 200,
      count: 1,
    }
    const tree = renderer.create(<Canvas {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
