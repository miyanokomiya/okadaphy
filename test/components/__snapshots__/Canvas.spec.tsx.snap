// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/ConfigForm snapshot 1`] = `
<div>
  <canvas
    height={200}
    width={100}
  />
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Pause
        </span>
      </button>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Step
        </span>
      </button>
    </div>
  </div>
</div>
`;
